### PR TITLE
[DEV-2727] fetch paymentInfo with receiver address

### PR DESCRIPTION
**Pull request by lapatric**
- [DEV-2727] support multi-address swaps

[DEV-2727]: https://dfxswiss.atlassian.net/browse/DEV-2727?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/react/src/hooks/api.hook.ts [hunk 1]
```diff
@@ -14,6 +14,7 @@ export interface CallConfig {
   data?: any;
   noJson?: boolean;
   specialHandling?: SpecialHandling;
+  headers?: Record<string, string>;
 }
```
Comment (davidleomay) [inline packages/react/src/hooks/api.hook.ts]: `token` field instead of `headers`?

File: packages/react/src/hooks/swap.hook.ts [hunk 2]
```diff
@@ -1,16 +1,28 @@
...
+  const { user } = useUserContext();
+
+  const accessToken = useRef<string | undefined>();
 
   async function receiveFor(info: SwapPaymentInfo): Promise<Swap> {
-    return call<Swap>({ url: SwapUrl.receive, method: 'PUT', data: info });
+    const request = { url: SwapUrl.receive, method: 'PUT', data: info } as CallConfig;
+    if (info.receiverAddress && user?.activeAddress?.address !== info.receiverAddress) {
+      accessToken.current ??= (await changeUserAddress(info.receiverAddress)).accessToken;
+      return call<Swap>({ ...request, headers: { Authorization: `Bearer ${accessToken.current}` } });
```
Comment (davidleomay) [inline packages/react/src/hooks/swap.hook.ts]: Reset access token on 401?

File: packages/react/src/hooks/api.hook.ts
Comment (davidleomay) [inline packages/react/src/hooks/api.hook.ts:1]: image
Don't reset the access token on 401, if a custom token was provided?

File: packages/react/src/contexts/session.context.tsx [hunk 3]
```diff
@@ -66,6 +70,7 @@ export function SessionContextProvider({ api, data, children }: SessionContextPr
   const [isProcessing, setIsProcessing] = useState(false);
   const [storedSignature, setStoredSignature] = useState<string>();
   const [storedAddress, setStoredAddress] = useState<string>();
+  const tokenCache = useRef<Map<string, string>>(new Map());
```
Comment (davidleomay) [inline packages/react/src/contexts/session.context.tsx:74]: Do you need the ref, or can you use a normal state?

**Review by davidleomay — APPROVED** [review]

File: packages/react/src/contexts/session.context.tsx (hunk 3 above)
Comment (lapatric) [reply packages/react/src/contexts/session.context.tsx:74]: I think in this case ref is more efficient as it avoids re-renders which shouldn't be necessary 🤔

**Review by davidleomay — APPROVED** [review]

**Pull request merged**